exercises/exercise_session_1.tex

```diff
@@ -0,0 +1,57 @@
+
+In this exercise session, we will study and apply some of these abstractions in order to captivate their importance, and to what end they can be utilized.
+
+\section{Exercises}
+\label{sec:exercises}
+
+\subsection*{Exercise 1}
+\label{sec:exercise_1}
+
+\emph{Define liveness and safety formally. Additionally, specify an example for both liveness and safety (the examples in the lecture and the book do not count).}
+
+\subsection*{Exercise 2}
+\label{sec:exercise_2}
+
+\emph{In class we saw that liveness and safety can be employed to prove the correctness of a distributed system. Let us assume a traditional producer-consumer setting. Which of the following are liveness properties and why?}
+
+\begin{enumerate}
+\item A consumer will eventually consume an element from the queue.
+\item Two consumers cannot consume the same head item concurrently.
+\item A producer has to stop producing if the queue is full.
+\item After a producer stops, it will start again if the queue is not full.
+\end{enumerate}
+
+\subsection*{Exercise 3}
+\label{sec:exercise_3}
+
+\emph{What is the problem vector clocks solved compared to logical clocks? How do they operate differently? In what type of timing abstraction are vector clocks employed?}
+
+\subsection*{Exercise 4}
+\label{sec:exercise_4}
+
+\emph{Let us assume we have a cluster where users can submit optimization procedures to a processing queue, e.g., for training neural networks. We know that every optimization process will be allocated to a single node. What kind of mechanisms do we need to construct in order to detect the crash of the process (or node), and insure that the neural network does not have to be trained from the start? Make a diagram, and write the procedure down in pseudo-code. Furthermore, identify the abstractions.}
+
+\subsection*{Exercise 5}
+\label{sec:exercise_5}
+
+\emph{We would like to process a very large dataset $\mathcal{D}$ (e.g., $> 4$ PB) of meteorological data to obtain the average temperature of a given location in a given time-frame. We define our computation in a acyclic graph structure $\mathcal{G}$ (drawn on the blackboard). This computation graph (this topic is explained in detail in later lectures) will now be submitted to the cluster, where all nodes will participate in the computation. However, a node fails during the computation. Can you come up with several approaches to recover from such a failure? What are the positive and negative aspects of every approach?}
+
+\subsection*{Exercise 6}
+\label{sec:exercise_6}
+
+\emph{Can you come up with a leader-election abstraction which is resilient against Byzantine adversaries, given the fact that the initial leader is correct? Proof informally (tip: Gibbard–Satterthwaite theorem).}
```

Comment:
About leader election and failure detection, you could also maybe reason with them regarding why they are equivalent.

Reply:
I'll put this one in my notes, thanks.